docs/developer/architecture/user_customization_and_persistence.rst

```diff
@@ -0,0 +1,225 @@
+    data:
+      config.yaml: |-
+        apiVersion: metalk8s.scality.com/v1alpha1
+        kind: GrafanaUserConfig
+        spec:
+          replicas: 2
+
+How it works
+~~~~~~~~~~~~
+
+Service pods and deployments will be configured to consume configuration data
+directly from their respective minion external pillars.
+
+During Bootstrap, these external pillar values will be pre-filled with default
+values and the service consumers will be configured to use these values.
+
+**Using Saltstates**
+
+Once a ConfigMap KV store is updated by the user(say a user changes the
+number of replicas for Prometheus deployments to a new value), then perform the
+following actions;
+
+  - Apply a salt state that reads the ConfigMap object, validates the schema
+    based on MetalK8s defined standards and checks the new values passed.
+  - If the ConfigMap object is valid, the new values passed by the user are
+    then re-rendered to the pillars.
+  - Finally, we make sure that the updated values are picked up by their
+    respective consumers(this might require Pod restarts for changes to take
+    effect).
+
+Note that, salt-states are used to sync data and update consumers
+of new configurations changes mainly because of the minimum effort it takes to
+setup this flow(i.e. the act of configuration update by the user and it's
+propagation to the consumers) but the K8s Operator pattern could be use to
+replace configuration synchronization between user defined configurations and
+consumers.
+
+The Operator approach is much more complex, requires much more effort
+to realize and there is no real need for applying changes using this method
+because configuration changes are not frequent(for a typical MetalK8s admin,
+changing the number of replicas for a given deployment could happen once in 3
+months) as such, having an operator watch for object changes is not significant
+and not very useful at this point in time.
+
+**Using Operator architecture(Custom Controllers)**
+
+When using an Operator(a Custom Controller that works with CRDs), we create a
+Custom Resource Definition (CRD) that references a ConfigMap. Once a ConfigMap
+is updated by the user, then the Operator is designed to perform the following
+actions;
+
+  - The Operator is connected to the API server to watch for changes in the
+    ConfigMap.
+  - If the Operator detects a modified ConfigMap, it then determines which line
+    of action it should take which are;
+
+      - Extract the ConfigMap name and object fields
+      - Extract the pods associated to this ConfigMap based on it's labels
+      - Read and validate the ConfigMap data and schema
+      - If the ConfigMap is valid, update the pillars and restart the
+        respective pods such that they pick-up new configs from the pillars.
+      - If the ConfigMap is invalid, log the error and perform no further
+        action because a bad ConfigMap being applied could lead to cluster
+        outages
```

Comment:
Hmm I not really like this point, we not directly apply the ConfigMap but we use it to modify another object and i don't see case where an invalid value can cause outages for me apiServer will just not set the wrong value.
For example if you use a string as replica then opeator will try to edit an object and apiServer will just reject the change, ... 
Maybe for other custom tuning that could lead to cluster outages but ... I don't see where

Reply:
ok... resolved!